bluetooth: sbc: move sbc to bluetooth

MarkWangChinese opened the pull request:
move sbc to bluetooth because only bluetooth uses it,
change CONFIG_LIBSBC_ENCODER and CONFIG_LIBSBC_DECODER as CONFIG_LIBSBC.

https://github.com/zephyrproject-rtos/zephyr/pull/95137#issuecomment-3288763900 suggest to move sbc to bluetooth.

MarkWangChinese committed:
bluetooth: sbc: move sbc to bluetooth

move sbc to bluetooth because only bluetooth uses it,
change CONFIG_LIBSBC_ENCODER and CONFIG_LIBSBC_DECODER as CONFIG_LIBSBC.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
libsbc: fix build issue

Remove `-Wno-array-bounds` as there does not seem to be any issues
that require it anymore.
Add -Wno-stringop-overflow to avoid what seem like false positive
warnings with gcc 13.3.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jhedberg commented on modules/libsbc/Kconfig:
```diff
@@ -4,12 +4,7 @@
-	bool "libsbc decoder Support"
-	help
-	  This option enables the Android SBC decoder library for Bluetooth A2DP
+	  This option enables low Complexity Subband Coding (SBC)
```
Shouldn't this be "low-complexity subband codec"? (based on https://en.wikipedia.org/wiki/SBC_(codec) )

MarkWangChinese replied:
In A2DP spec, The AAC is `Advanced Audio Coding` too. And I ask the question to AI, the answer is:

> Coding refers to the specific algorithm or method used to encode/decode data:
It's the mathematical process and technique
Describes the encoding scheme itself
More academically precise term

> Codec refers to the implementation (software/hardware) that performs coding:
It's the actual encoder/decoder device or program
More of a product or system term
Combines "coder" + "decoder"

So, it seems that the `Codec` is more appropriate as you said. Let me change it.

jhedberg reviewed (changes requested):
What's the reason for this move? The commit messages don't explain that?

MarkWangChinese commented:
Hi @jhedberg Could you please help to review this pr again? Thanks.

jhedberg commented:
> In native_sim platform building: the __packed leads building errors

@MarkWangChinese what exactly are the errors? This is important, so we know that you're not hiding something that should really be fixed in the code.

MarkWangChinese commented:
> @MarkWangChinese what exactly are the errors? This is important, so we know that you're not hiding something that should really be fixed in the code.

Building cmd: west build -b native_sim/native ./tests/bluetooth/shell -T bluetooth.shell.shell_br -d .\buildwsl\test1 -p always
The error is as follow:
```
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/include/zephyr/bluetooth/sbc.h:18:27: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or __attribute__’ before ‘{’ token
   18 | enum __packed sbc_ch_mode {
      |                           ^
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/include/zephyr/bluetooth/sbc.h:30:30: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or __attribute__’ before ‘{’ token
   30 | enum __packed sbc_alloc_mthd {
      |                              ^
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/include/zephyr/bluetooth/sbc.h:56:29: error: field ‘alloc_mthd’ has incomplete type
   56 |         enum sbc_alloc_mthd alloc_mthd;
      |                             ^~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/include/zephyr/bluetooth/sbc.h:58:26: error: field ‘ch_mode’ has incomplete type
   58 |         enum sbc_ch_mode ch_mode;
      |                          ^~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c: In function ‘sbc_frame_bytes’:
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:104:64: error: ‘SBC_CH_MODE_MONO’ undeclared (first use in this function)
  104 |                 (encoder->sbc_encoder_params.s16ChannelMode == SBC_CH_MODE_MONO ? 1 : 2);
      |                                                                ^~~~~~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:104:64: note: each undeclared identifier is reported only once for each function it appears in
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c: In function ‘sbc_frame_encoded_bytes’:
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:119:47: error: ‘SBC_CH_MODE_MONO’ undeclared (first use in this function)
  119 |         if (encoder_params->s16ChannelMode == SBC_CH_MODE_MONO) {
      |                                               ^~~~~~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:125:48: error: ‘SBC_CH_MODE_DUAL_CHANNEL’ undeclared (first use in this function); did you mean ‘SBC_DUAL_CHANNEL’?
  125 |             (encoder_params->s16ChannelMode == SBC_CH_MODE_DUAL_CHANNEL)) {
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
      |                                                SBC_DUAL_CHANNEL
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:128:54: error: ‘SBC_CH_MODE_STEREO’ undeclared (first use in this function)
  128 |         } else if (encoder_params->s16ChannelMode == SBC_CH_MODE_STEREO) {
      |                                                      ^~~~~~~~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c: In function ‘sbc_frame_bytes’:
/mnt/c/_DDM/zephyr/zephyrproject3/zephyr/subsys/bluetooth/lib/sbc.c:105:1: warning: control reaches end of non-void function [-Wreturn-type]
  105 | }
      | ^
[44/354] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/shell/modules/kernel_service/kernel_shell.c.obj
ninja: build stopped: subcommand failed.
```
From my see, the `__packed` is not recognized by the compiler. Note: I am using the WSL of windows to build it.
After added `#include <zephyr/toolchain.h>`, there are no errors.

jhedberg commented:
> After added #include <zephyr/toolchain.h>, there are no errors.

Wouldn't that be the better fix then?

MarkWangChinese commented:
@jhedberg  For `-Wno-stringop-overflow`.
The native_sim platform building warnings are as follow, I don't see the building warnings on one NXP cortex-m7 platform (mimxrt1170_evk@B/mimxrt1176/cm7). The codes are from Android bluetooth SBC codec.
```
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
In file included from /mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc_private.h:74,
                 from /mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_bitstream.h:40,
                 from /mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:37:
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset 16 into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset 17 into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [18, 34] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [19, 35] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [20, 36] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [21, 37] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [22, 38] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [23, 39] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [24, 40] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [25, 41] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [26, 42] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [27, 43] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:141:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  141 |       *scale_factor++ = HIGH(f = *b++);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [28, 44] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/srce/decoder-private.c:142:23: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |       *scale_factor++ = LOW(f);
      |                       ^
/mnt/c/_DDM/zephyr/zephyrproject3/modules/lib/libsbc/decoder/include/oi_codec_sbc.h:197:10: note: at offset [29, 45] into destination object ‘scale_factor’ of size 16
  197 |   int8_t scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];
      |          ^~~~~~~~~~~~
[26/28] Linking C executable zephyr/zephyr.elf
```

The `scale_factor[SBC_MAX_CHANNELS * SBC_MAX_BANDS];` is defined [here](https://github.com/zephyrproject-rtos/libsbc/blob/main/decoder/include/oi_codec_sbc.h#L197). (SBC_MAX_CHANNELS is 2 and SBC_MAX_BANDS is 8 from spec, so the `scale_factor` is 16 bytes. The above warnings happen [here](https://github.com/zephyrproject-rtos/libsbc/blob/main/decoder/srce/decoder-private.c#L129-L143). Actually the `i` will not exceed the 16 bytes because `common->frameInfo.nrof_subbands` is not bigger than SBC_MAX_BANDS and `common->frameInfo.nrof_channels` is not bigger than SBC_MAX_CHANNELS. But the native_sim compiler report this warning.

jhedberg commented:
@aescolar any thoughts on the above? FWIW I don't think this has as much to do with native_sim as it has to do with what local toolchain you have on your host machine.

MarkWangChinese commented:
> @MarkWangChinese I can see the issue with gcc 13.3.0, but not with 14.2.0. Given that and your explanation, I'm guessing it is likely a false positive in gcc 13.3. So I think it is fine to disable these warnings. But I would explain it as such in modules/libsbc/CMakeLists.txt. Something like:
> 
> ```
> # -Wno-stringop-overflow was added as otherwise gcc 13.3 produces multiple warnings in this library
> # Newer version of gcc do not produce them, and this code is taken from Android where it was
> # validated. So these warnings are presumed to be false positives.
> ```
> 
> Also please fix the first commit msg to something like
> 
> ```
> Include `zephyr/toolchain.h` explicitly as it is needed by this header,
> instead of relaying on it being included indirectly by others.
> Remove `-Wno-array-bounds` as there does not seem to be any issues
> that require it anymore.
> Add -Wno-stringop-overflow to avoid what seem like false positive
> warnings with gcc 13.3.
> ```

Thanks very much @aescolar I have updated the pr as your suggestions.

kartben reviewed (changes requested):
@MarkWangChinese thanks for taking care of this!
As this might impact users downstream (need to update include path) given this is a public header, could you please add a quick note in doc/releases/migration-guide-4.4.rst ? thanks!

MarkWangChinese committed:
doc: releases: migration-guide-4.4: libsbc is moved under bluetooth

Document that libsbc (sbc.c and sbc.h) has been moved under the Bluetooth
subsystem, with sbc.h now located in include/zephyr/bluetooth.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented:
> @MarkWangChinese thanks for taking care of this! As this might impact users downstream (need to update include path) given this is a public header, could you please add a quick note in doc/releases/migration-guide-4.4.rst ? thanks!

Hi @kartben I add one commit to update the doc, please help to review the commit. Thanks.

MarkWangChinese committed:
libsbc: remove unused compile definitions

They are used in the sbc of bluedroid not in the sbc of Android bluetooth,
and Zephyr is using the sbc of Android bluetooth, so remove them.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jhedberg commented:
Looks like this caused a [regression in CI](https://github.com/zephyrproject-rtos/zephyr/actions/runs/19925705596/job/57124841314). @MarkWangChinese could you please look at fixing it ASAP?
```
[181/354] Building C object modules/libsbc/CMakeFiles/libsbc.dir/__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c.obj
FAILED: modules/libsbc/CMakeFiles/libsbc.dir/__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c.obj 
ccache /usr/lib/llvm-20/bin/clang -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DMBEDTLS_CONFIG_FILE=\"config-mbedtls.h\" -DZVFS_OPEN_SIZE=0 -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/twister-out/native_sim_native/llvm/tests/bluetooth/shell/bluetooth.shell.shell_br/zephyr/include/generated/zephyr -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/native_sim_native/llvm/tests/bluetooth/shell/bluetooth.shell.shell_br/zephyr/include/generated -I/__w/zephyr/zephyr/soc/native/inf_clock -I/__w/zephyr/zephyr/boards/native/native_sim -I/__w/zephyr/zephyr/lib/midi2/. -I/__w/zephyr/zephyr/scripts/native_simulator/common/src/include -I/__w/zephyr/zephyr/scripts/native_simulator/native/src/include -I/__w/zephyr/zephyr/subsys/testsuite/include -I/__w/zephyr/zephyr/subsys/bluetooth -I/__w/zephyr/zephyr/subsys/shell/modules/kernel_service/thread/.. -I/__w/zephyr/modules/hal/microchip/include -I/__w/zephyr/modules/hal/st/sensor/inc -I/__w/zephyr/modules/hal/ti/mspm0/source/ti/devices/msp/. -I/__w/zephyr/modules/hal/ti/mspm0/source/ti/devices/msp/m0p -I/__w/zephyr/modules/hal/ti/mspm0/source/ti/devices/msp/peripherals -I/__w/zephyr/modules/hal/ti/mspm0/source/ti/devices/msp/peripherals/m0p -I/__w/zephyr/modules/hal/ti/mspm0/source/ti/devices/msp/peripherals/m0p/sysctl -I/__w/zephyr/modules/lib/libsbc/encoder/include -I/__w/zephyr/modules/lib/libsbc/decoder/include -I/__w/zephyr/modules/lib/libsbc/decoder/srce -I/__w/zephyr/modules/crypto/mbedtls/include -I/__w/zephyr/zephyr/modules/mbedtls/configs -I/__w/zephyr/zephyr/modules/mbedtls/include -fno-strict-aliasing -Werror -Oz -imacros /__w/zephyr/zephyr/twister-out/native_sim_native/llvm/tests/bluetooth/shell/bluetooth.shell.shell_br/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics --config=/__w/zephyr/zephyr/cmake/toolchain/llvm/clang_libgcc.cfg -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-unused-but-set-variable -Wno-typedef-redefinition -Wno-deprecated-non-prototype -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wno-self-assign -Wno-initializer-overrides -Wno-section -Wno-gnu -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/__w/zephyr/zephyr/tests/bluetooth/shell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -msse2 -mfpmath=sse -fvisibility=hidden -std=c17 -O3 -std=c11 -ffast-math -Wno-stringop-overflow -MD -MT modules/libsbc/CMakeFiles/libsbc.dir/__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c.obj -MF modules/libsbc/CMakeFiles/libsbc.dir/__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c.obj.d -o modules/libsbc/CMakeFiles/libsbc.dir/__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c.obj -c /__w/zephyr/modules/lib/libsbc/encoder/srce/sbc_analysis.c
error: unknown warning option '-Wno-stringop-overflow'; did you mean '-Wno-shift-overflow'? [-Werror,-Wunknown-warning-option]```
